client/clientservice/src/client_service.cpp

```diff
@@ -15,22 +15,66 @@
 
 namespace concord::client::clientservice {
 
-void ClientService::start(const std::string& addr, uint64_t max_receive_msg_size) {
+void ClientService::start(const std::string& addr, int num_async_threads, uint64_t max_receive_msg_size) {
   grpc::EnableDefaultHealthCheckService(true);
 
   grpc::ServerBuilder builder;
   builder.AddListeningPort(addr, grpc::InsecureServerCredentials());
   builder.SetMaxReceiveMessageSize(max_receive_msg_size);
-  builder.RegisterService(request_service_.get());
+
+  // Register synchronous services
   builder.RegisterService(event_service_.get());
 
+  // Register asynchronous services
+  builder.RegisterService(&request_service_);
+
+  for (int i = 0; i < num_async_threads; ++i) {
+    cqs_.emplace_back(builder.AddCompletionQueue());
+  }
+
   auto clientservice_server = std::unique_ptr<grpc::Server>(builder.BuildAndStart());
 
+  // From the "C++ Performance Notes" in the gRPC documentation:
+  // "Right now, the best performance trade-off is having numcpu's threads and one completion queue per thread."
+  for (int i = 0; i < num_async_threads; ++i) {
+    server_threads_.emplace_back(std::thread([this, i] { this->handleRpcs(i); }));
+  }
+
   auto health = clientservice_server->GetHealthCheckService();
   health->SetServingStatus(kRequestService, true);
   health->SetServingStatus(kEventService, true);
 
   clientservice_server->Wait();
+
+  LOG_INFO(logger_, "Wait for async gRPC threads to finish");
+  for (auto& t : server_threads_) {
+    t.join();
+  }
+}
+
+void ClientService::handleRpcs(int thread_idx) {
+  // Note: Memory is freed in `proceed`
+  new requestservice::RequestServiceCallData(&request_service_, cqs_[thread_idx].get(), client_);
+
+  void* tag;  // uniquely identifies a request.
```

Comment:
Nit: I think the tag can be of type `requestservice::RequestServiceCallData*` and then implicitly converted to `void*`, avoiding the static_cast below.

Reply:
This is from the example code and I prefer to leave it that way. It emphasizes that the completion queue just deals with `void*` which can represent anything. This will later be used if we decide to make more services asynchronous. In addition, to my knowledge the static_cast doesn't add any performance overhead.